File: 0162-find-peak-element/0162-find-peak-element.cpp
class Solution {
public:
    int findPeakElement(vector<int>& nums) {
        int start = 0;
        int n = nums.size();
        int end = n - 1;
        int mid = start + ((end- start)/2) ;
        while(start < end){ //start < end hoga peak element find karne me
            if(nums[mid] < nums[mid +1]){
                start = mid + 1;
            }
            else {
                end = mid;
            }
            mid = start + ((end - start)/2) ;
        }
        return start;
    }
};

File: 0162-find-peak-element/0162-find-peak-element.py
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        left = 0 
        right = len(nums) - 1
        

        while(left < right):
            mid = (left + right)//2

            if nums[mid] < nums[mid + 1]:
                left = mid + 1

            else:
                right =  mid  
                
        return left




# nums[mid] < nums[mid + 1]
# The slope is going up → a peak must exist to the right.

# nums[mid] > nums[mid + 1]
# The slope is going down → a peak exists at mid or to the left.


#  Time complexity O(log n)